### DISCUSSION TITLE
GenericFunctionMaterial: add and multiply?

### DISCUSSION THREAD

**Discussion by jmeier** [idaholab/moose#28027] [Q&A General] [answered]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

Dear Moose-Community,

With this post, I would like to discuss with you whether we should add four new (optional) parameters to [GenericFunctionMaterial](https://mooseframework.inl.gov/source/materials/GenericFunctionMaterial.html). I am somewhat undecided here: On the one hand, it would make input files shorter and easier to read and require fewer function objects. On the other hand, it would double functionality.

**Initial situation**
For models in which different material parameters have to be multiplied by the same factor, several functions often have to be defined per material parameter and per element group (e.g. subdomain). In the simplest case, two functions per parameter and subdomain: at least one function with the "original" (non-multiplied) distribution and one for the multiplied distribution. The same applies in the case of addition (parameter value offset). 

**Possible Solution**
The definition of the second function could be omitted if [GenericFunctionMaterial](https://mooseframework.inl.gov/source/materials/GenericFunctionMaterial.html) could take over this multiplication/addition directly. 

This could look like this in the input file:
```
[./gfm]
  type = GenericFunctionMaterial
  block = 'ThisBlockOverThere'
  prop_names = density
  prop_values = dens_func
  multiply_by_const = 3.9  # constant values to multiply prop_values by (before add)
  multiply_by = 'MyRampFunction1'  # functions to multiply prop_values by (before add)
  add_const = 85.7  # constant values to add to prop_values (after multiplication)
  add = 'MyRampFunction2'  # functions to add to prop_values (after multiplication)
[]
```

Advantages:
- Fewer functions, as the addition/multiplication takes place directly in [GenericFunctionMaterial](https://mooseframework.inl.gov/source/materials/GenericFunctionMaterial.html)
- Shorter input files (easier to read, reduced potential for input errors)

Disadvantages:
- Duplication of functionality (the [GenericFunctionMaterial](https://mooseframework.inl.gov/source/materials/GenericFunctionMaterial.html) can now calculate)
- More complexity in [GenericFunctionMaterial](https://mooseframework.inl.gov/source/materials/GenericFunctionMaterial.html)

Please let me know what you think about this extension to the GenericFunctionMaterial. If this addition makes sense from your point of view, I can try to program it and then submit a PR.

Jörg

**Comment by GiudGiud** [comment]
Hello

No these four hard-coded operations wont be needed
Can you use a ParsedMaterial preferrably?
If it does not support all you need then we'll need to add that

Guillaume

**Reply by jmeier** [reply]
Dear Guillaume,

Thank you for your answer and ponting to [ParsedMaterial](https://mooseframework.inl.gov/source/materials/ParsedMaterial.html).
As good as ParsedMaterial looks to me, it doesn't seem to be able to multiply by another function value.

In the spirit of your answer I would like to explore all possibilities before proposing new code:

Question 1:
As far I understood ParsedMaterial, it can use variables (coupled_variables) and a user-defined FParser-function. But is there a way to use another Moose-Function as input?

Question 2:
I know I can use `t` to make the function in ParsedMaterial dependent on time. But is there a way to make a Moose variable dependent on time? 
This is not working: `bla = '${fparse (34 + t)*0}'`

Jörg

**Reply by GiudGiud** [reply]
Hello

Ok I was semi expecting this.
ParsedMaterial doesn't support functions directly. 
However it supports auxiliary variables, and ParsedAux supports functions through the use of the functor_names arguments

You should be able to forward things that way into a material property

It would be good to simplify this if you have development time

Guillaume

**Reply by jmeier** [reply]
Dear Guillaume,

In this case, perhaps I should adapt my request? Wouldn't it make more sense to extend ParsedMaterial with the option to access Moose functions and leave GenericFunctionMaterial as it is? (Disclaimer: I haven't looked at the code of ParsedMaterial yet. I have no idea whether this suggestion makes sense).

Jörg

**Reply by GiudGiud** [reply, marked as answer]
This makes sense. 
Happy to support that. 

You may want to add functor support instead, this adds support for functions, variables, pps and functor materials. 

And this list is bound to expand

**Reply by jmeier** [reply]
Now we have a whish list! When I have time, I'll have a look through the code. I'll close this thread due to fact, the initial question has been answerd. Once again, thanks for your support!

**Reply by GiudGiud** [reply]
https://mooseframework.inl.gov/source/materials/MaterialFunctorConverter.html this should be a good example to look at to create a functor argument
then ParsedAux should show how to add functors to the parsed expression, with all the nice checks

**Reply by jmeier** [reply]
I'm now looking into extending ParsedMaterial with functor support. Here I need some advice: My first approach was to add the parameters "functor_names" and "functor_symbols" to `ParsedMaterialBase`. While adding the parameters itself is not a probem, I like to check them and call `paramError `in case something is not right. Due to the fact, `ParsedMaterialBase` seems not to derive from `MooseObject`, I cannot call `paramError `. Shall I move these two parameters to `ParsedMaterialHelper` (what feels "wrong")? Or shall I make ParsedMaterialBase derive from MooseObject? Or I'm overlooking something?

**Reply by GiudGiud** [reply]
You can use a pointer to the object (in the code: this), which you can cast as well, to call paramError

**Reply by jmeier** [reply]
I'm sorry, I need more help.

`MooseObject * mo_ptr = dynamic_cast<MooseObject *>(this);`

gives an compiler error:

`error: cannot 'dynamic_cast' '(ParsedMaterialBase*)this' (of type 'class ParsedMaterialBase*') to type 'class MooseObject*' (source type is not polymorphic)`

**Comment by GiudGiud** [comment]
ah too bad.

Ok in the constructor of ParsedMaterialBase, could you start passing a pointer to the class which is building it?
Is ParsedMaterialBase only created as a base class of classes that derive from MooseObjects?  This would let you do a constructor like:

ParsedMaterialBase::ParsedMaterialBase(const InputParameters & parameters, MooseObject * derived_class)

then we could do something like
```
_derived_class->paramError(
```

**Reply by jmeier** [reply]
Thank you! This is how I implemented this now. 

While advancing with my code, I hit some more problems. I would like to coordinate how best to proceed here. Is it better to ask my questions here in this discussion, or should I better create a PR (as a draft?). Obviosly, this PR will initially be incomplete.

**Reply by GiudGiud** [reply]
Draft PR is a good idea

**Reply by jmeier** [reply]
Issue: #28220
PR (draft for now): #28221
